solidity/contracts/ECDSARewardsDistributor.sol

```diff
@@ -0,0 +1,96 @@
+  ▓▓▓▓▓▓▄▄▓▓▓▓▓▓▓▀      ▐▓▓▓▓▓▓▄▄▄▄         ▓▓▓▓▓▓▄▄▄▄         ▐▓▓▓▓▓▌   ▐▓▓▓▓▓▓
+  ▓▓▓▓▓▓▓▓▓▓▓▓▓▀        ▐▓▓▓▓▓▓▓▓▓▓▌        ▓▓▓▓▓▓▓▓▓▓▌        ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓
+  ▓▓▓▓▓▓▀▀▓▓▓▓▓▓▄       ▐▓▓▓▓▓▓▀▀▀▀         ▓▓▓▓▓▓▀▀▀▀         ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▀
+  ▓▓▓▓▓▓   ▀▓▓▓▓▓▓▄     ▐▓▓▓▓▓▓     ▓▓▓▓▓   ▓▓▓▓▓▓     ▓▓▓▓▓   ▐▓▓▓▓▓▌
+▓▓▓▓▓▓▓▓▓▓ █▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓ ▐▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓  ▓▓▓▓▓▓▓▓▓▓
+
+                           Trust math, not hardware.
+*/
+
+pragma solidity 0.5.17;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "@keep-network/keep-core/contracts/utils/BytesLib.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/SafeERC20.sol";
+import "@keep-network/keep-core/contracts/KeepToken.sol";
+import "@openzeppelin/contracts/cryptography/MerkleProof.sol";
+
+contract ECDSARewardsDistributor is Ownable {
+    using SafeERC20 for KeepToken;
+    using BytesLib for bytes;
+
+    KeepToken public token;
+
+    // This event is triggered whenever a call to #claim succeeds.
+    event Claimed(uint256 index, address account, uint256 amount);
+
+    event MerkleRootAdded(bytes32 merkleRoot);
+
+    // merkleRoot -> total amount for distribution for a given week
+    mapping(bytes32 => uint256) private merkleRoots;
+
+    // This is a packed array of booleans.
+    mapping(bytes32 => mapping(uint256 => uint256)) private claimedBitMap;
+
+    constructor(address _token) public {
+        token = KeepToken(_token);
+    }
+
+    /// Call receiveApproval to "allocate" amount of KEEP for a given merkleRoot
+    /// @param _extraData merkle root (32 bytes)
+    function receiveApproval(
+        address _from,
+        uint256 _amount,
+        address _token,
+        bytes memory _extraData
+    ) public {
+        require(IERC20(_token) == token, "Unsupported token");
+
+        // TODO: check _amount
+        token.safeTransferFrom(_from, address(this), _amount);
+
+        // TODO: check _extraData length
+        bytes32 merkleRoot = _extraData.toBytes32();
+
+        merkleRoots[merkleRoot] = _amount;
+
+        emit MerkleRootAdded(merkleRoot);
+    }
+
+    function isClaimed(uint256 index, bytes32 merkleRoot) public view returns (bool) {
+        uint256 claimedWordIndex = index / 256;
+        uint256 claimedBitIndex = index % 256;
+        uint256 claimedWord = claimedBitMap[merkleRoot][claimedWordIndex];
+        uint256 mask = (1 << claimedBitIndex);
+        return claimedWord & mask == mask;
+    }
+
+    function _setClaimed(uint256 index, bytes32 merkleRoot) private {
+        uint256 claimedWordIndex = index / 256;
+        uint256 claimedBitIndex = index % 256;
+        claimedBitMap[merkleRoot][claimedWordIndex] = claimedBitMap[merkleRoot][claimedWordIndex] | (1 << claimedBitIndex);
+    }
+
+    function claim(uint256 index, address account, uint256 amount, bytes32 merkleRoot, bytes32[] calldata merkleProof) external {
```

Comment:
I believe there is a vulnerability here! We are not checking whether the user-provided `merkleRoot` is defined and what is the allocation for it. I think an adversary can drain funds from this contract only if they provide a valid root and proof for their inputs.

Reply:
Not sure if I'm following. Merkle root is generated using the list of stakers `<addresses : KEEP_amount>`. If someone will call this function providing a proper `merkleRoot` & `merkleProofs` but invalid `amount`, we would revert with `Invalid proof` error. We also have a check for if the rewards were already claimed for the given staker.

Reply:
`merkleRoot` and `merkleProof` are passed by the user. The user can generate any input of their choice including a tree that will include proof of a reward for their account. We do not check whether the merkle tree passed as a parameter has an allocation.

Reply:
In other words, the fact we have `onlyOwner` on `receiveApproval` does not help if the user can pass any `merkleRoot` and we dot validate if there is an allocation for this root.

Reply:
Good catch! https://github.com/keep-network/keep-ecdsa/pull/627/commits/733e67d023178a7a1ad23f4eb358f8a6851db3a4